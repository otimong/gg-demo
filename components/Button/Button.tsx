import React, { Component } from 'react';
import { Icon } from '../Icon'
import classNames from 'classnames';


export interface ButtonProps {
    icon?: string

    iconPosition?: 'left' | 'right'
	/**
	 * Add more style classes to the component
	 */
    propClasses?: Array<string>
	/**
	 * If provided, will turn the button into a link 
	 */
    href?: string
	/**
	 * Font scale. A bigger scale means small text. 6 is smaller than 1 
	 */
    fontScale?: number
	/**
	 * An object with css in-line styles 
	 */
    style?: React.CSSProperties

	/**
	 * When you click on the button
	 */
    onClick?: () => void

    size?: 'small' | 'medium' | 'large'

    type?: 'primary' | 'secondary'
}

export class Button extends Component<ButtonProps, Object> {

    static defaultProps: ButtonProps = {
        size: 'medium',
        type: 'primary'
    }

    render() {
        const { propClasses, href, type, style, fontScale = 6, children, onClick, icon, iconPosition, size, ...others } = this.props;
        const buttonClasses = classNames(
            'p-2',
            'rounded',
            {
                'py-1': size === 'small',
                'py-2': size === 'medium',
                'py-4': size === 'large',
                'bg-blue': type === 'primary',
                'text-white': type === 'primary',
                'bg-yellow': type === 'secondary'
            },
            propClasses
        );
        const iconClasses = classNames(
            'ph2',
            {
                'fr': iconPosition === 'right',
                'fl': iconPosition === 'left'
            })

        const buttonContent = typeof children === 'string' ? children.toString() : 'Button Text';

        let button = (
            <button {...others} className={buttonClasses} onClick={onClick}>
                {icon && <Icon type={icon} classes={iconClasses} />}
                {buttonContent}
            </button>
        );

        return href
            ?
            <a {...others} className={buttonClasses} href={href}>{buttonContent}</a>
            :
            button;
    }
}