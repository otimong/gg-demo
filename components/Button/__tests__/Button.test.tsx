import React from 'react'
import { shallow, configure } from 'enzyme'

import { Button } from '../Button'

describe("Button", () => {
    it("Should render correctly", () => {
        let wrapper = shallow(<Button />)
        expect(wrapper).toMatchSnapshot()
    })

    it("Should render an icon", () => {
        let wrapper = shallow(<Button icon="twitter" />)
        expect(wrapper).toMatchSnapshot()
    })
})