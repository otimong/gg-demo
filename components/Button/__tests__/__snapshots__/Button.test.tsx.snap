// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Should render an icon 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    icon="twitter"
    size="medium"
    type="primary"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Icon
          classes="ph2"
          size="lg"
          type="twitter"
        />,
        "Button Text",
      ],
      "className": "p-2 rounded py-2 bg-blue text-white",
      "onClick": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "classes": "ph2",
          "size": "lg",
          "type": "twitter",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "Button Text",
    ],
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Icon
            classes="ph2"
            size="lg"
            type="twitter"
          />,
          "Button Text",
        ],
        "className": "p-2 rounded py-2 bg-blue text-white",
        "onClick": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "classes": "ph2",
            "size": "lg",
            "type": "twitter",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "Button Text",
      ],
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Button Should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    size="medium"
    type="primary"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        undefined,
        "Button Text",
      ],
      "className": "p-2 rounded py-2 bg-blue text-white",
      "onClick": undefined,
    },
    "ref": null,
    "rendered": Array [
      undefined,
      "Button Text",
    ],
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          undefined,
          "Button Text",
        ],
        "className": "p-2 rounded py-2 bg-blue text-white",
        "onClick": undefined,
      },
      "ref": null,
      "rendered": Array [
        undefined,
        "Button Text",
      ],
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
